src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        # input query [ngram, T,B,C]
+        # input attn_weights [ngram, B*head,T,S]
+        # input real_positions [B,T] or [1,1]
+        # input i_bucket_relative_stream [B,T, 2*T] or None
+
+        N, T, B, _ = query.size()
+        _, BH, _, S = attn_weights.size()
+
+        if i_bucket_relative_stream is not None:
+            i_buckets = i_bucket_relative_stream
+        else:
+            # [B,T,S]
+            assert real_positions[0][0] == S - 1, 'memory position is 1 2 3 4 5(S-1)'
+            relative_positions = torch.arange(0, S).unsqueeze(0).unsqueeze(0).repeat(B, T, 1).to(real_positions.device)
+            # [B,T,1]
+            real_positions = real_positions.unsqueeze(0).repeat(B, T, 1)
+            relative_positions = relative_positions
+            # [B,T,2*T] or [B,T,S]
+            relative_positions = relative_positions - real_positions
+            i_buckets = self._relative_positions_bucket(relative_positions, False)
+
+        # [ngram, B, T, C]
+        query = query.transpose(1, 2)
+        # [ngram, B, T, bucket*head]
+        values = self.relative_linear(query)
+        # [ngram, B, T, bucket, head]
+        values = values.view(*values.size()[:-1], self.num_buckets, self.num_heads)
+        # [ngram, B, head, T, bucket]
+        values = values.permute(0, 1, 4, 2, 3)
+        # [ngram*B*head, T, bucket]
+        values = values.reshape(N * BH, T, -1)
+
+        # [ngram, B, head*T, S]
+        i_buckets = i_buckets.unsqueeze(0).repeat(N, 1, self.num_heads, 1)
+
+        values = values.reshape(-1, values.size(-1))
+        i_buckets = i_buckets.view(-1, i_buckets.size(-1)).long()
+        # [ngram*B*head*T, S]
+        result = torch.gather(values, dim=1, index=i_buckets)
+        # [ngram, B*head, T, S]
+        result = result.view(N, BH, T, -1)
+
+        return result
+
+    def forward(self, query, key, value, key_padding_mask=None, layer_state=None,
+                need_weights=True, static_kv=False,
+                self_attn_mask=None,
+                ngram_mask_matrix=None,
+                i_buckets_main_stream=None,
+                i_bucket_relative_stream=None,
+                real_positions=None,
+                output_attentions = False
+                ):
+
+        tgt_len, bsz, embed_dim = query.size()
+        assert embed_dim == self.embed_dim
+        assert list(query.size()) == [tgt_len, bsz, embed_dim]
+
+        if layer_state is not None:  # reuse k,v and encoder_padding_mask
+            saved_state = layer_state.get(self.cache_key, {})
+            if 'prev_key' in saved_state:
+                # previous time steps are cached - no need to recompute
+                # key and value if they are static
+                if static_kv:
+                    assert self.encoder_decoder_attention and not self.self_attention
+                    key = value = None
+        else:
+            saved_state = None
+            layer_state = {}
+
```

Comment:
If I read the function correctly the function arguments `key` and `value` are not used -> can we delete them from the function signature?